src/java/org/apache/cassandra/service/TruncateResponseHandler.java

```diff
@@ -61,24 +63,34 @@ public TruncateResponseHandler(int responseCount)
     public void get() throws TimeoutException
     {
         long timeoutNanos = getTruncateRpcTimeout(NANOSECONDS) - (nanoTime() - start);
-        boolean completedInTime;
+        boolean signaled;
         try
         {
-            completedInTime = condition.await(timeoutNanos, NANOSECONDS); // TODO truncate needs a much longer timeout
+            signaled = condition.await(timeoutNanos, NANOSECONDS); // TODO truncate needs a much longer timeout
         }
         catch (InterruptedException e)
         {
             throw new UncheckedInterruptedException(e);
         }
 
-        if (!completedInTime)
-        {
+        if (!signaled)
             throw new TimeoutException("Truncate timed out - received only " + responses.get() + " responses");
-        }
 
-        if (truncateFailingReplica != null)
+        if (!failureReasonByEndpoint.isEmpty())
         {
-            throw new TruncateException("Truncate failed on replica " + truncateFailingReplica);
+            // clone to make sure no race condition happens
+            Map<InetAddressAndPort, RequestFailureReason> failureReasonByEndpoint = new HashMap<>(this.failureReasonByEndpoint);
+            int size = failureReasonByEndpoint.size();
+            long timeouts = failureReasonByEndpoint.values().stream().filter(RequestFailureReason.TIMEOUT::equals).count();
+            long nonTimeout = size - timeouts;
+            if (nonTimeout <= timeouts)
+                throw new TimeoutException("Truncate timed out - received only " + responses.get() + " responses");
+
+            StringBuilder sb = new StringBuilder("Truncate failed on ");
+            for (Map.Entry<InetAddressAndPort, RequestFailureReason> e : failureReasonByEndpoint.entrySet())
+                sb.append("replica ").append(e.getKey()).append(' ').append(e.getValue()).append(", ");
```

Comment:
nit: maybe a color after the address would be helpful?
```suggestion
                sb.append("replica ").append(e.getKey()).append(': ').append(e.getValue()).append(", ");
```

Reply:
`:` may be an issue as that could imply port is missing... but we could do ` -> ` instead of the space?  I also want to make sure there is a space between the replica to avoid any existing log parsers freaking out.

Reply:
`->` WFM